### ISSUE TITLE
Wrong no-useless-escape when escaping character after escaping "\"

### ISSUE THREAD

**Issue by MarkusLund** [eslint/eslint#12565]
**Tell us about your environment**

Node version: v10.7.0
npm version: v6.8.0
Local ESLint version: v6.6.0
Global ESLint version: v6.6.0 (Currently used)

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

The problem exists with default settings. 

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

The problem is when you try to escape a character after escaping a backslash.
```js
const regEx =  "[\\\/]"; //match backslash or forwardslash
```

Command used:
```bash
eslint
```

**What did you expect to happen?**
No eslint error/warning. 

**What actually happened? Please include the actual, raw output from ESLint.**
no-useless-escape error/warning. 

EDIT: Changed the example. Exclamation does not need an escape. But forwardslash does.

**Comment by mdjermanovic** [comment]
Hi @MarkusLund, thanks for the issue!

The example indeed looks like a useless escape:

```js
"[\\\!]" === "[\\!]" // true
```

**Comment by aladdin-add** [comment]
https://xkcd.com/1638/

**Comment by MarkusLund** [comment]
Oh.. I changed the example when trying different things. It should escape both backslash and forwardslash. I Updated the issue.

**Comment by aladdin-add** [comment]
![image](https://user-images.githubusercontent.com/13050025/68935878-4d537e00-07d4-11ea-81b9-64781f3e1be5.png)

there is no need to escape `"/"` in plain string(but required in regex)

**Comment by mdjermanovic** [comment]
@MarkusLund if the example is meant to be a part of a dynamically generated regex pattern, in particular a part to match either backslash or forward-slash (as described in the comment), then it should be something like this:

```js
const regExPart =  "[\\\\/]"; //match backslash or forwardslash
```

or, more readable with `String.raw`:

```js
const regExPart =  String.raw`[\\/]`; //match backslash or forwardslash
```

There is no need to escape the forward-slash when the regex is created from a string:

```js
new RegExp("/").test("/"); // true
```

If that's unexpected and confusing, you can also keep the escape there:

```js
const regExPart =  "[\\\\\\/]"; //match backslash or forwardslash

const regExPart =  String.raw`[\\\/]`; //match backslash or forwardslash
```

The last example is exactly how that part would look in a regex literal:

```js
/[\\\/]/
```

If the regex isn't dynamically generated, it seems better to use regex literals and avoid possible mistakes with escape sequences, as described in the [prefer-regex-literals](https://eslint.org/docs/rules/prefer-regex-literals) rule.

The `"[\\\/]"` example from the original post would be `\[\/]\` in a regex literal, and matches only the forward-slash:

```js
// same as "[\\/]", same effect in new RegExp() as just "[/]", equivalent to /[\/]/
const regEx =  "[\\\/]"; 

new RegExp(regEx).test("/"); // true
new RegExp(regEx).test("\\"); // false

console.log(regEx); // [\/]
```

**Comment by MarkusLund** [comment]
Thanks for clarifying, but my initial example should have been a literal, and not a string. 

I get the same warning when I do 
```js
const regex = /[\\\/]/;
```

Using `eslint:recommended` I still get the warning `no-useless-escape` when using a regex literal.

**Comment by mysticatea** [comment]
The `\/` is a useless escape in character classes.

- `/[\/]/` equals `/[/]/`

Of course, the backslash is needed outside of character classes.

- `/\//`

**Comment by MarkusLund** [comment]
Ok, thanks for taking you time clarifying.
